wandb/sdk/launch/create_job.py

```diff
@@ -0,0 +1,430 @@
+    if entrypoint:
+        if not os.path.exists(os.path.join(local_dir, entrypoint)):
+            wandb.termerror(f"Invalid entrypoint: {entrypoint}")
+            return None
+        # prepend working dir (specified in path)
+        entrypoint = os.path.join(src_dir, entrypoint)
+    elif not filename:
+        if path.endswith(".py"):
+            wandb.termerror("Invalid entrypoint provided in path")
+        else:
+            wandb.termerror(
+                "Invalid entrypoint. Provide an entrypoint in the path or use -E"
+            )
+        return None
+    else:  # try to make one from the path, assumes working dir is parent of file
+        rel_entrypoint = os.path.join(src_dir, filename)
+        if rel_entrypoint.startswith("/"):
+            rel_entrypoint = rel_entrypoint[1:]
+
+        if not os.path.exists(os.path.join(tempdir, rel_entrypoint)):
+            wandb.termerror(
+                f"Entrypoint {rel_entrypoint} not valid, specify one in the path or use -E"
+            )
+            return None
+        entrypoint = rel_entrypoint
+
+    # check if requirements.txt exists
+    if not os.path.exists(os.path.join(local_dir, "requirements.txt")):
+        repo_formd = path.replace(entrypoint, "")
+        wandb.termerror(
+            f"Could not find requirements.txt file in git repo at: {repo_formd}/requirements.txt"
+        )
+        return None
+
+    metadata = {
+        "git": {
+            "commit": commit,
+            "remote": ref.url,
+        },
+        "root": ref.repo,
+        "codePath": entrypoint,
+        "entrypoint": [f"python{python_version}", entrypoint],
+        "python": python_version,  # used to build container
+    }
+
+    return metadata
+
+
+def _create_artifact_metadata(
+    path: str, entrypoint: str, runtime: Optional[str] = None
+) -> Tuple[Dict[str, Any], List[str]]:
+    if not os.path.exists(path):
+        wandb.termerror("Path must be a valid file or directory")
+        return {}, []
+
+    if not os.path.exists(os.path.join(path, "requirements.txt")):
+        wandb.termerror(f"Could not find requirements.txt file in: {path}")
+        return {}, []
+
+    # read local requirements.txt and dump to temp dir for builder
+    requirements = []
+    with open(os.path.join(path, "requirements.txt")) as f:
+        requirements = f.read().splitlines()
+
+    if runtime:
+        python_version = _clean_python_version(runtime)
+    else:
+        python_version = ".".join(get_current_python_version())
+
+    metadata = {"python": python_version, "codePath": entrypoint}
+    return metadata, requirements
+
+
+def _handle_artifact_entrypoint(
+    path: str, entrypoint: Optional[str] = None
+) -> Tuple[str, Optional[str]]:
+    if os.path.isfile(path):
+        if entrypoint and path.split("/")[-1] != entrypoint:
+            wandb.termwarn(
+                "Ignoring passed in entrypoint, as 'path' is to a file, which will be used as the entrypoint"
```

Comment:
nit: This warning is confusing. I think it can be improved by indicating that the user is supplying path, and where the user is supplying path

Reply:
Also, if the file is nested this check will fail if the user passes an entrypoint with multiple directories

Reply:
What does "entrypoint with multiple directories" look like?

Reply:
jk that makes sense

Reply:
Although because we don't know what the root dir is, at least in this artifact case, we just assume that the root is the parent of the direct filepath. Not ideal, and certainly wrong in some cases, but I don't think we can do better?

The reason to support this default would be to make basic use-cases super slick. For most practitioners, it feels likely that they might create jobs from a pretty simple project folder, so supporting commands like `wandb job create code main.py` feels best.

Reply:
A user should be able to say:
`wandb create code ./my/path/to/dir -E from/dir/to/code.py`
no?

Reply:
<img width="1014" alt="Screen Shot 2023-07-06 at 9 26 02 AM" src="https://github.com/wandb/wandb/assets/19414170/79c03887-2f0b-4211-a112-73bba0ce0a5c">
yeah that works

Reply:
How do I use this to handle this case:
```
main_dir
│
├── subdir1
│   ├── subdir2
│   │   └── train.py
│   
└── lib
```

Where I want the root to be `main_dir` but I want the entrypoint to be train.py. I can't figure it out

Reply:
<img width="940" alt="Screen Shot 2023-07-06 at 10 20 51 AM" src="https://github.com/wandb/wandb/assets/19414170/3edc267d-7487-41a7-b6fb-9a147872ed2d">

Either with path or entrypoint?

Reply:
<img width="1101" alt="Screen Shot 2023-07-06 at 10 23 55 AM" src="https://github.com/wandb/wandb/assets/19414170/551cb8f3-7878-4b06-be8c-b85be44d1dbe">
Both of these produce runnable jobs, but with different names, which is probably a bug

Reply:
What does **not** work would be to omit the entrypoint entirely: 
<img width="921" alt="Screen Shot 2023-07-06 at 10 29 07 AM" src="https://github.com/wandb/wandb/assets/19414170/e04d9ab1-66e8-4e87-86f4-c69cfc668cc4">
@KyleGoyette

Reply:
nit: I was having issues because we were following a dead symlink in a wandb dir. I foresee this being a somewhat common problem if I ran into it twice. We should make sure we capture errors on `add_dir` and display them nicely.